src/cli/filter_arg.zig

```diff
@@ -0,0 +1,195 @@
+const strings = bun.strings;
+const json_parser = bun.JSON;
+const Glob = @import("../glob.zig");
+
+const Package = @import("../install/lockfile.zig").Package;
+
+fn findWorkspaceMembers(allocator: std.mem.Allocator, log: *bun.logger.Log, workspace_map: *Package.WorkspaceMap, workdir_: []const u8) !void {
+    bun.JSAst.Expr.Data.Store.create(bun.default_allocator);
+    bun.JSAst.Stmt.Data.Store.create(bun.default_allocator);
+
+    defer {
+        bun.JSAst.Expr.Data.Store.reset();
+        bun.JSAst.Stmt.Data.Store.reset();
+    }
+
+    var workdir = workdir_;
+
+    while (true) : (workdir = std.fs.path.dirname(workdir) orelse break) {
+        const parent_trimmed = strings.withoutTrailingSlash(workdir);
+        var name_buf: bun.PathBuffer = undefined;
+        @memcpy(name_buf[0..parent_trimmed.len], parent_trimmed);
+        name_buf[parent_trimmed.len..name_buf.len][0.."/package.json".len].* = "/package.json".*;
+        name_buf[parent_trimmed.len + "/package.json".len] = 0;
+        const json_path = name_buf[0 .. parent_trimmed.len + "/package.json".len];
+
+        log.msgs.clearRetainingCapacity();
+        log.errors = 0;
+        log.warnings = 0;
+
+        const json_file = std.fs.cwd().openFileZ(
+            name_buf[0 .. parent_trimmed.len + "/package.json".len :0].ptr,
+            .{ .mode = .read_only },
+        ) catch continue;
+        defer json_file.close();
+
+        const json_stat_size = try json_file.getEndPos();
+        const json_buf = try allocator.alloc(u8, json_stat_size + 64);
+        defer allocator.free(json_buf);
+        const json_len = try json_file.preadAll(json_buf, 0);
+        const json_source = bun.logger.Source.initPathString(json_path, json_buf[0..json_len]);
+        const json = try json_parser.ParseJSONUTF8(&json_source, log, allocator);
+
+        const prop = json.asProperty("workspaces") orelse continue;
+
+        const json_array = switch (prop.expr.data) {
+            .e_array => |arr| arr,
+            .e_object => |obj| if (obj.get("packages")) |packages| switch (packages.data) {
+                .e_array => |arr| arr,
+                else => break,
+            } else break,
+            else => break,
+        };
+        const name_prop = json.asProperty("name") orelse continue;
+        const name = switch (name_prop.expr.data) {
+            .e_string => |n| n.data,
+            else => break,
+        };
+
+        _ = Package.processWorkspaceNamesArray(
+            workspace_map,
+            allocator,
+            log,
+            json_array,
+            &json_source,
+            prop.loc,
+            null,
+        ) catch |err| {
+            return err;
+        };
+
+        // add the root package to the workspace map, too
+        const entry = Package.WorkspaceMap.Entry{ .name = try allocator.dupe(u8, name), .version = null, .name_loc = bun.logger.Loc.Empty };
+        try workspace_map.insert(try allocator.dupe(u8, parent_trimmed), entry);
+
```

Comment:
This duplicates memory twice.

Reply:
Does it? From my understanding we dupe the name (part of the hash map value) in the first line, and duple the path (hash map key) in the second line.

Reply:
Right here:

https://github.com/oven-sh/bun/blob/52894121ea2436481b97b5318d1ce64176b18c0e/src/install/lockfile.zig#L3562-L3573